lib/ecto/type.ex

```diff
@@ -916,75 +914,78 @@ defmodule Ecto.Type do
   """
   @spec equal?(t, term, term) :: boolean
   def equal?(type, term1, term2) do
-    if requires_semantic_comparison?(type) do
-      do_equal?(type, term1, term2)
+    if fun = equal_fun(type) do
+      fun.(term1, term2)
     else
       term1 == term2
     end
   end
 
-  defp requires_semantic_comparison?(:decimal),
-    do: true
-  defp requires_semantic_comparison?(:date),
-    do: false
-  defp requires_semantic_comparison?(type) when type in @calendar,
-    do: true
-  defp requires_semantic_comparison?(type) when type in @base,
-    do: false
-  defp requires_semantic_comparison?({:array, type}),
-    do: requires_semantic_comparison?(type)
-  defp requires_semantic_comparison?({:map, type}),
-    do: requires_semantic_comparison?(type)
-  defp requires_semantic_comparison?(module) when is_atom(module),
-    do: loaded_and_exported?(module, :equal?, 2)
+  defp equal_fun(:decimal), do: &equal_decimal?/2
+  defp equal_fun(t) when t in [:time, :time_usec], do: &equal_time?/2
+  defp equal_fun(t) when t in [:utc_datetime, :utc_datetime_usec], do: &equal_utc_datetime?/2
+  defp equal_fun(t) when t in [:naive_datetime, :naive_datetime_usec], do: &equal_naive_datetime?/2
+  defp equal_fun(t) when t in @base, do: nil
 
-  defp do_equal?(:decimal, %Decimal{} = term1, %Decimal{} = term2) do
-    Decimal.equal?(term1, term2)
-  end
-  defp do_equal?(type, %Time{} = term1, %Time{} = term2) when type in [:time, :time_usec] do
-    Time.compare(term1, term2) == :eq
-  end
-  defp do_equal?(type, %NaiveDateTime{} = term1, %NaiveDateTime{} = term2)
-       when type in [:naive_datetime, :naive_datetime_usec] do
-    NaiveDateTime.compare(term1, term2) == :eq
-  end
-  defp do_equal?(type, %DateTime{} = term1, %DateTime{} = term2)
-       when type in [:utc_datetime, :utc_datetime_usec] do
-    DateTime.compare(term1, term2) == :eq
-  end
-  defp do_equal?(type, term1, term2) when type in @base do
-    term1 == term2
+  defp equal_fun({:array, type}) do
+    if fun = equal_fun(type) do
```

Comment:
@josevalim the biggest gain is here, we only call this function "once" (its recursive) instead of per element in the list which is important for custom types that require function_exported? check.

We were discussing that we should perform similar optimisation for load/dump/cast but I don't think it's necessary; in the `cast/2` call here https://github.com/elixir-ecto/ecto/blob/9f8ddb3da57305d526e5fba845e3700297a78f25/lib/ecto/type.ex#L613 we could already know that it's say :decimal so we could have some `cast_decimal` but letting beam find matching clause of cast/2 should be fast enough right? Maybe I misunderstood.

Reply:
@wojtekmach we will do the exact same changes for load/dump/cast. Since we will make the callback optional, we will know in many situations that we can just return `{:ok, value}` without traversing at all. For example, for an array of integers, we don't need to traverse it either, we just need to return `{:ok, value}`.